File: algorithms/medium/matchsticks_to_square.ex
defmodule Solution do
  use Bitwise

  @spec makesquare(matchsticks :: [integer]) :: boolean
  def makesquare(matchsticks) do
    {sum, len} =
      matchsticks
      |> Enum.reduce({0, 0}, fn m, {sum, len} ->
        {sum + m, len + 1}
      end)

    if rem(sum, 4) != 0 do
      false
    else
      target = div(sum, 4)

      matchsticks = matchsticks |> Enum.with_index()

      mask_end = (1 <<< len) - 1

      0..mask_end
      |> Enum.reduce(%{0 => 0}, fn mask, dp ->
        if dp[mask] == nil do
          dp
        else
          matchsticks
          |> Enum.reduce(dp, fn {m, i}, dp ->
            j = 1 <<< i

            if (mask &&& j) == 0 && dp[mask] + m <= target do
              Map.put(dp, mask ||| j, rem(dp[mask] + m, target))
            else
              dp
            end
          end)
        end
      end)
      |> Map.get(mask_end)
      |> then(&(&1 == 0))
    end
  end
end

File: algorithms/medium/matchsticks_to_square.rs
impl Solution {
    pub fn makesquare(matchsticks: Vec<i32>) -> bool {
        let sum: i32 = matchsticks.iter().sum();
        if sum % 4 != 0 {
            return false;
        }
        let target = sum / 4;
        let mut dp = vec![-1; 1 << matchsticks.len()];
        dp[0] = 0;
        for mask in 0..dp.len() {
            if dp[mask] == -1 {
                continue;
            }

            for (i, m) in matchsticks.iter().enumerate() {
                let j = 1 << i;
                if mask & j == 0 && dp[mask] + m <= target {
                    dp[mask | j] = (dp[mask] + m) % target;
                }
            }
        }
        dp.pop().unwrap() == 0
    }
}
